src/contexts/WalletContext/useWeb3Auth/index.ts

```diff
@@ -1,112 +1,140 @@
-      console.error(error);
-    } finally {
-      return web3Auth;
-    }
+  const [state, setState] = useState<WalletState>({ status: "disconnected" });
+
+  // //// EVENT HANDLERS ////
+  const handleChainChange: ChainChangeHandler = (hexChainId) => {
+    setState((p) =>
+      p.status === "connected"
+        ? { ...p, chainId: new Decimal(hexChainId).toString() }
+        : p
+    );
   };
+  const handleAccountsChange: AccountChangeHandler = (accounts) => {
+    setState((p) => {
+      if (p.status !== "connected") return p;
+      if (isEmpty(accounts)) {
+        //side effects that don't modify state
+        p.provider.removeListener("chainChanged", handleChainChange);
+        p.provider.removeListener("accountsChanged", handleAccountsChange);
+        web3Auth.logout({ cleanup: true });
 
+        return { status: "disconnected" };
+      }
+
+      return { ...p, address: accounts[0] };
+    });
+  };
+
+  // //// PERSISTENT CONNECTION ////
   useEffect(() => {
-    const setupProviderInfo = async () => {
+    const adapter = web3Auth.cachedAdapter;
+    if (adapter) login(false);
+    //eslint-disable-next-line
+  }, []);
+
+  // //// USER INITIATED CONNECTION ////
+  const login = async (isNew = true) => {
+    try {
+      setState({ status: "loading" });
+
+      //there's always a need to initialize Web3Auth
+      await web3Auth.init();
+
+      const provider = isNew
+        ? await web3Auth.connectTo("torus-evm")
+        : web3Auth.provider;
+
       if (!provider) {
-        console.log("provider not initialized yet");
-        return;
+        // don't throw error on persistent connection
+        if (isNew) return;
+        throw new Error("Failed to connect to wallet");
+      }
+
+      const accounts = await provider.request<string[]>({
+        method: EIPMethods.eth_requestAccounts,
+      });
+      const hexChainId = await provider.request<string>({
+        method: EIPMethods.eth_chainId,
+      });
+
+      if (!(val(accounts) && val(hexChainId))) {
+        if (isNew) return;
+        throw new Error("Failed to connect to wallet");
       }
-      const rpc = new RPC(provider);
-      const chainId = await rpc.getChainId();
-      const address = await rpc.getAccounts();
-      setProviderInfo({
-        providerId: "web3auth-metamask",
-        logo: WEB3AUTH_LOGO,
-        chainId,
-        address,
+
+      provider.on("chainChanged", handleChainChange);
+      provider.on("accountsChanged", handleAccountsChange);
+
+      setState({
+        status: "connected",
+        address: accounts[0],
+        chainId: new Decimal(hexChainId).toString(),
+        provider: provider,
       });
-    };
-    if (provider) setupProviderInfo();
-  }, [provider, chainId]);
-
-  const login = async () => {
-    setIsLoading(true);
-    const web3Auth = await init();
-    if (!web3Auth) {
-      console.log("web3auth not initialized yet");
-      return;
+    } catch (err) {
+      logger.error(err);
```

Comment:
What kind of errors can appear here? Should they some error modal be shown to the user?